keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/LogProbGradientCalculator.java

```diff
@@ -0,0 +1,157 @@
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Probabilistic;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.Differentiator;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+
+public class LogProbGradientCalculator {
+
+    private final Set<? extends Vertex<?>> logProbOfVertices;
+    private final Set<? extends Vertex<?>> wrtVertices;
+
+    private final Map<Vertex, Set<DoubleVertex>> parentToLatentLookup;
+    private final Map<Vertex, Set<DoubleVertex>> verticesWithNonzeroDiffWrtLatent;
+
+    public LogProbGradientCalculator(List<? extends Vertex> logProbOfVerticesList, List<? extends Vertex<?>> wrtVerticesList) {
+        this.logProbOfVertices = new HashSet<>((List<Vertex<?>>) logProbOfVerticesList);
+        this.wrtVertices = new HashSet<>(wrtVerticesList);
+
+        parentToLatentLookup = getParentsThatAreConnectedToWrtVertices(logProbOfVertices);
+        verticesWithNonzeroDiffWrtLatent = getVerticesWithNonzeroDiffWrt(logProbOfVertices, parentToLatentLookup);
+    }
+
+    /**
+     * @return the partial derivatives with respect to a given set of latent vertices
+     */
+    public Map<VertexId, DoubleTensor> getJointLogProbGradientWrtLatents() {
+        PartialDerivatives diffOfLogWrt = new PartialDerivatives(new HashMap<>());
+
+        for (final Vertex<?> ofVertex : logProbOfVertices) {
+            diffOfLogWrt = diffOfLogWrt.add(reverseModeLogProbGradientWrtLatents(ofVertex));
+        }
+
+        return diffOfLogWrt.asMap();
+    }
+
+    /**
+     * The dLogProb(x) method on Vertex returns a partial derivative of the Log Prob with respect to each
+     * of its arguments and with respect to its value, x. This method searches these partials for any that
+     * are parents of the vertices we are taking the derivative with respect to
+     *
+     * @param ofVertices          the vertices that the derivative is being calculated "of" with respect to the wrtVertices
+     * @param parentToWrtVertices a lookup
+     * @return a map for a given vertex to a set of the wrt vertices that it is connected to
+     */
+    private Map<Vertex, Set<DoubleVertex>> getVerticesWithNonzeroDiffWrt(Set<? extends Vertex<?>> ofVertices, Map<Vertex, Set<DoubleVertex>> parentToWrtVertices) {
+        return ofVertices.stream()
+            .collect(Collectors.toMap(
+                v -> v,
+                v -> {
+                    Set<DoubleVertex> parents = v.getParents().stream()
+                        .map(parent -> (DoubleVertex) parent)
```

Comment:
It's non-zero because it's connected to `vertices that we are taking the derivative with respect to.` If it's not connected to one of these then the derivative with respect to those will always be zero.